absolute-quantum/cats-blender-plugin
# Rewrite of shapekey.py for performance and to maintain relative keys and vertex groups

I have entirely rewritten shapekey.py (apply active shape key to basis) to significantly increase its speed and to maintain shape keys' relative keys and vertex groups.

Bugs with the previous implementation that are fixed or no longer relevant:
- Confusing behaviour caused by basis' relative key being set to the reverted key after application
- Issue preventing reverting more than one applied key
- Missing check for object mode in poll()

Improvements over previous implementation:
- Allows keys to be reverted in any order
- Maintains shape key order
- Maintains relative key of all shape keys
  - The difference between all shape keys and their relative keys remain the same (except the reverted key by design)
- Maintains vertex groups on all shape keys (including the applied shape key)
- Maintains active shape key
- Significant performance improvements, see benchmarking results for different numbers of shape keys and vertices: https://docs.google.com/spreadsheets/d/1PABGG2EztNAlBVwDPT5M-pOQz_YaYB3npRAQVfXDq5I/edit?usp=sharing
  - Admittedly, when the mesh has upwards of about 100K vertices, a large amount of time in the original implementation will probably be spent on going into edit mode and changing the selection and doing 'remove doubles', but I have no idea what causes the bug that's supposed to fix (shape key explosion) and I've not had that bug happen since I stopped using Cats buttons to separate and join meshes over a year ago. The only current known shapekey related blender issues I could find and replicate are https://developer.blender.org/T44415 and an issue with curves and surfaces https://developer.blender.org/T78701.

-----

The tests run.py script has been modified to allow for separate sets of tests to be run against separate test .blend files in different folders. All of the tests relevant to the armature .blend files have been moved to their own folder as the tests and download location for the .blend file for shapekey.py have been put in their own folder too. All of the test changes could be removed or significantly changed if necessary. I made sure that running the run.py test script with all default arguments would result in the same behaviour as before (running the armature tests against the armature blend files).

I couldn't get the viseme or translate tests to ever pass (would always give me the "It looks like you got banned from google translate temporarily"), though I haven't tried with any recent changes to the development branch (not that the tests themselves have changed at all).

Test changes:
- Remove hardcoded requirement for tests and blend files to be in "./tests"
- Add details on how to run the tests (top of run.py, could put it somewhere else)
- Replace boolean test option arguments with argumentless options
- Add tests option -e to continuing running further tests even if a test fails
- Add tests option -p to pipe all output to stdout and stderr instead of using custom filtering and output, mainly for debugging tests
- Add ModuleNotFoundError to checks for invalid test
- Remove no longer existing mmd_tools from list of addons to load during tests
- Generalise unit tests for any test blend files instead of hardcoding for armature blend files (-a and --armature options renamed to -f and --bfile respectively)
- Reorganise tests into separate folders based on the test blend files and the tests that go with them

-----
I did try extending shapekey.py to all object types that support shape keys, but curves in particular are problematic. While a curve's shape key can be activated with a negative value to make a control point's radius go below zero, the shape key itself has a hard minimum value of zero. Tilt also has a hard maximum and minimum that could be conceivably be reached. I could not think of any decent solution for these issues, so opted to leave the operator to only support meshes.
